### DISCUSSION TITLE
how can I get around `Target#apply() call cannot use`....

### DISCUSSION THREAD

**Discussion by axaluss** [com-lihaoyi/mill#1478] [Q&A]
"Target#apply() call cannot use `" + banned0.symbol + "` defined within the T{...} block"

How am i supposed to work around this?
I want to compose several command calls into one command.

Some wiki entry describing this or a better error message would really help.

**Comment by lefou** [comment]
Can you show a snippet of your target and the full error message.

**Comment by axaluss** [comment]
Weird that `m.publish` works but `_.scalaVersionShort()` does not...

```scala
def publishAll(
      scalaVersion: String,
      sonatypeCreds: String,
      release: Boolean
    ): Command[Unit] =
    T.command {
      mainModule
        .millInternal
        .modules
        .collect {
          case m: PublishModule with Base =>
            m
        }
        .filter(_.scalaVersionShort() == scalaVersion)
        .foreach { m =>
          m.publish(sonatypeCreds, release = release, stagingRelease = false)
        }
    }
class Base  extends extends ScalaModule with PublishModule {
//.....
  override def scalaVersion: T[String] = T { scalaVersionLong }
}
```

```
build.sc:109: Target#apply() call cannot use `value x$12` defined within the T{...} block
        .filter(_.scalaVersion() == scalaVersion)
                ^
Compilation Failed
```

**Comment by lefou** [comment]
Sorry, I still don't understand. I can't see the definition of `scalaVersionShort()`. Also your snippet seems unrelated to the shown error message. I can't find the code from the error message anywhere.

**Comment by lefou** [comment]
Maybe your are looking for one of the combinators `T.zip`, `T.traverse` or `T.sequence`?

**Comment by axaluss** [comment]
```scala
def publishAll(
      scalaVersion: String,
      sonatypeCreds: String,
      release: Boolean
    ): Command[Unit] =
    T.command {
      mainModule
        .millInternal
        .modules
        .collect {
          case m: PublishModule with Base =>
            m
        }
        .filter(_.scalaVersion() == scalaVersion)
        .foreach { m =>
          m.publish(sonatypeCreds, release = release, stagingRelease = false)
        }
    }
class Base(_scalaVersionShort: String) extends ScalaModule with PublishModule {
//  val _scalaVersionShort: String         = ???
  override def scalaVersion: T[String] = T { _scalaVersionShort }
}
```

result:

```
build.sc:109: Target#apply() call cannot use `value x$12` defined within the T{...} block
        .filter(_.scalaVersion() == scalaVersion)
                ^
Compilation Failed
```


I then created a different version with a non `T` signature `def scalaVersion: String= _scalaVersionShort` 
and changed the filter to `.filter(_.scalaVersion == scalaVersion)`
in which the call to  `m.publish(...` works....

Why is the call to  `_.scalaVersion() ` a problem, but `m.publish(...` is not?

**Reply by lefou** [reply]
In your example you are collecting all `PublishMoudle`s. After that you want to evaluate a target (`scalaVersion`) in these modules. And with the result, you want to evaluate another target (`publish`) for some modules. This kind of imperative evaluations isn't supported in a single run. But it is possible to realize this with a [Evaluator Commands](https://com-lihaoyi.github.io/mill/mill/Extending_Mill.html#_evaluator_commands_experimental). With the evaluator at hand (injected as the first parameter), you can start multiple evaluations. Have have look at some other places where evaluator commands are used, e.g. `GenIdea/idea`.

**Comment by lefou** [comment]
It's probably because you are never executing the resulting `publish` target. I think you are just constructing a publish command in the `foreach` and then drop it. But you never call `.apply()` on in.

**Comment by axaluss** [comment]
Ah ok so both tasks `scalaVersion` and `publish` would be failing to compile if i would try to compose them. 

Some chapter about working around `call cannot use` might be helpful for newcomers.

And I have to use a combination of T.zip, T.traverse or T.sequence instead to ship around the macro limitations?

**Comment by axaluss** [comment]
How could I compose existing commands and tasks etc into a new command In this example?

**Reply by lefou** [reply]
Se my [other comment](https://github.com/com-lihaoyi/mill/discussions/1478?converting=1#discussioncomment-1346034)
